[Improvement][Master] Add Worker Group validator in api side.

DarkAssassinator opened the pull request:
<!--Thanks very much for contributing to Apache DolphinScheduler. Please review https://dolphinscheduler.apache.org/en-us/community/development/pull-request.html before opening a pull request.-->

## Purpose of the pull request

fix #12001 

## Brief change log

Just add a validator before task instance was put to dispatch queue.
Why need this change?
Because check the worker group before add to dispatch queue can avoid invalid infinite loops in `TaskPriorityQueueConsumer`

## Verify this pull request

DarkAssassinator committed:
[Improvement][Master] Process instance will keep running when cannot find the worker group. #12001

DarkAssassinator committed:
Merge branch 'dev' into improve_12001

rickchengx committed:
[Improvement-11291][sql] Improvement the log output while executing multiple sql statements (#11292)

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/CommonTaskProcessor.java:
```diff
@@ -126,6 +133,14 @@ public boolean dispatchTask() {
             }
 
+            // check in advance to avoid invalid infinite loops in TaskPriorityQueueConsumer
+            if (CollectionUtils.isEmpty(serverNodeManager.getWorkerGroupNodes(taskExecutionContext.getWorkerGroup()))) {
```
We cannot make this change easily, the workgroup may loss in some cases, and recovery automically.

And this is not safe, the workflow may exist in the processor but loss in dispatch part.

BTY, it's not suggested to do this validation master, we need to do the validation in api-side.

DarkAssassinator replied:
> We cannot make this change easily, the workgroup may loss in some cases, and recovery automically.
> 
> And this is not safe, the workflow may exist in the processor but loss in dispatch part.
> 
> BTY, it's not suggested to do this validation master, we need to do the validation in api-side.

but if ds want to worker group recovery automically, we also should not add this validation in api-side,just keep loop waiting. WDYT @ruanwenjun

ruanwenjun replied:
Add validation in API-Side can stop the process in workflow creation. In fact, we have this check in front-end, but in api-side, we don't validate the workgroup.

DarkAssassinator replied:
> Add validation in API-Side can stop the process in workflow creation. In fact, we have this check in front-end, but in api-side, we don't validate the workgroup.

u are correct, i add a validator in api side and remove it in master side

fengjian1129 committed:
[Bug-#12057][task-plugin] fix when the sql query result is empty, the email fails to send the attachment, and an exception will be reported (#12059)

* [Bug-#12057][task-plugin] fix when the sql query result is empty, the email fails to send the attachment, and an exception will be reported

Co-authored-by: 冯剑 Jian <jian.feng@jiduauto.com>

DarkAssassinator committed:
add worker group validator in api side and format some lines

DarkAssassinator committed:
add worker group validator in api side and format some lines

DarkAssassinator committed:
add check process instance worker group when re-run

DarkAssassinator committed:
add check process instance worker group when re-run

DarkAssassinator committed:
add check process instance worker group when re-run

caishunfeng commented:
Hi @DarkAssassinator please use the correct title, thanks~

DarkAssassinator commented:
> Hi @DarkAssassinator please use the correct title, thanks~

ohh. so sorry. done

ruanwenjun commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ExecutorServiceImpl.java:
```diff
@@ -306,11 +311,64 @@ public Map<String, Object> checkProcessDefinitionValid(long projectCode, Process
+                    return result;
+                }
+                // check all sub process recursively
+                Map<String, Object> subResult = checkWorkerGroupNameExists(subProcessDefinition,
```
I am not sure if this will exist a cycle here.

DarkAssassinator replied:
> I am not sure if this will exist a cycle here.

because workflow is a DAG, so this change like a DFS, as i think it will not exist a cycle

DarkAssassinator committed:
Merge branch 'dev' into improve_12001

zhongjiajie committed:
[Feature][API] New restful API for workflow and schedule  (#11912)

* [feat] New restful API for workflow and schedule

CURD for workflow and schedule, different with exists
API, this new restful api only operate single resource
in each request, and return the latest. For example,
previous workflow should also need to post tasks definition
and tasks relation definition, but this patch will allow
you to create workflow without task relate information

* use checkProjectAndAuthThrowException, and fix CI error

* Update dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ScheduleV2Controller.java

* change method name
from createProcessDefinitionV2 to createSingleProcessDefinition
from updateProcessDefinitionV2 to updateSingleProcessDefinition

Co-authored-by: caishunfeng <caishunfeng2021@gmail.com>

EricGao888 committed:
[Doc][Improvement] Add instructions for process execution type and data complement (#12121)

* Add instructions for process execution type and data complement

* Fix and polish the docs

DarkAssassinator committed:
[fix#12000]Cannot remove the WorkerGroup from the master service (#12050)

* [Bug] [Master] Cannot remove the WorkerGroup from the master service. #12000

* remove unnecessary locks

* Update dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/registry/ServerNodeManager.java

Co-authored-by: caishunfeng <caishunfeng2021@gmail.com>

Co-authored-by: caishunfeng <caishunfeng2021@gmail.com>

DarkAssassinator committed:
update lambda

DarkAssassinator committed:
resolve conflicts

DarkAssassinator commented:
Misuse coverage in conflict resolution， close it and re-open
